Hiding Watched Videos on Channel View

palharesf opened the pull request:
# Issue 4497 - Hiding Watched Videos on Channel View

## Pull Request Type
- [ ] Bugfix
- [x] Feature Implementation
- [ ] Documentation
- [ ] Other

## Related issue
Closes #4497

## Description
The 'Hide Videos on Watch' setting has been moved from the 'Subscription' section to the 'Distraction Free' section. Now, it not only hides watched videos (and shorts and lives) from the subscription page, but also from the channel page (from the videos, shorts and lives tab, but not from the home tab).

## Screenshots
Current implementation:
![image](https://github.com/user-attachments/assets/648d94f4-46e0-4893-9475-560b56c90bd8)
![image](https://github.com/user-attachments/assets/e9204da9-7343-499a-8c1a-8bedad14e672)
![image](https://github.com/user-attachments/assets/c5d0e83d-4efb-403e-b24e-84e9eea6b39a)

Proposed change (settings page):
![image](https://github.com/user-attachments/assets/1c2c57a8-a865-44ec-879c-955dbdc19f87)
![image](https://github.com/user-attachments/assets/36b3076b-c783-47bb-b8ff-982a4b4cb207)

Channel pages (untoggled):
![image](https://github.com/user-attachments/assets/4836db4f-f558-4910-8680-58756478a565)
![image](https://github.com/user-attachments/assets/61bbd5ab-d708-4103-9f0b-07c1fe6889f8)
![image](https://github.com/user-attachments/assets/73a63b93-7613-4001-ab72-82794eefac1b)

Channel pages (toggled):
![image](https://github.com/user-attachments/assets/0dcab828-586e-42c2-b202-c74c82887eef)
![image](https://github.com/user-attachments/assets/d93cf2ca-e9e5-44a5-bb45-79f0e534600a)
![image](https://github.com/user-attachments/assets/4a700b7c-247a-4f62-a429-2746b5c7fc05)

Tooltip:
![image](https://github.com/user-attachments/assets/d7fc166a-df94-484d-9431-ef550a91624c)


## Testing
(A)
- Navigate to a random channel
- Navigate to Videos tab
- Watch a video or manually mark it as watched -- it should only gray out
- Go to Settings --> Distraction Free --> General --> toggle 'Hide Videos on Watch'
- Return to the previously channel page
- Watched videos should not load anymore
- Untoggling the 'Hide Videos on Watch' and returning to the page should reload the videos

(B)
- Repeat all steps from the test (A), but choose a channel with few videos
- Leave two unwatched videos
- Ensure the toggle is on
- Return to the channel page
- Upon marking one of the last two videos as watched, the 'Sort By' dropdown will disappear (since there's no need to sort a single video, shorts or live)
- Screenshots:
![image](https://github.com/user-attachments/assets/eab15f7d-f99c-4cc9-ae6b-497278f04066)
![image](https://github.com/user-attachments/assets/5284a803-be25-4286-a009-1935123ec0bd)


## Desktop
Windows 10 Pro
OS Build:	19045.5796
Freetube v0.23.4 Beta

## Additional context

efb4f5ff-1298-471a-8973-3d47447115dc commented:
hi @palharesf in the testing section you mainly are mentioning how you tested it. It would be nice if you could provide some testcases (ofcourse we will be testing outside of those cases too but would be nice to have some kind of baseline)

palharesf commented:
Oh, of course @efb4f5ff-1298-471a-8973-3d47447115dc - I thought the request was for information on how I tested it hahaha

For a specific Channels page, the feature is differentiating between four kinds of states:
- No videos watched, toggle off: all videos should be visible
- No videos watched, toggle on: all videos should be visible
- One or more videos watched, toggle: all videos should be visible
- One or more videos watched, toggle on: one or few videos should not be visible

If I were to write an automated test for this, I would implement these four test cases.

Is that what you were alluding to, @efb4f5ff-1298-471a-8973-3d47447115dc ?

efb4f5ff-1298-471a-8973-3d47447115dc reviewed (commented):
The Shorts and the Live tab also contains videos that can marked as watched so it would be a more consistent user experience to also include the toggle on these tabs

palharesf commented:
I didn't originally touch Shorts and Live as I felt those were outside the scope of the feature, but I'm happy to incorporate those changes tomorrow to harmonize the experience. I'll work on that tomorrow unless told otherwise

efb4f5ff-1298-471a-8973-3d47447115dc commented:
> Oh, of course @efb4f5ff-1298-471a-8973-3d47447115dc - I thought the request was for information on how I tested it hahaha
 
Maybe that section has to be revised so its more clear that those need to be provided, ill look into it.

> For a specific Channels page, the feature is differentiating between four kinds of states:
> - No videos watched, toggle off: all videos should be visible
> - No videos watched, toggle on: all videos should be visible
> - One or more videos watched, toggle: all videos should be visible
> - One or more videos watched, toggle on: one or few videos should not be visible
> If I were to write an automated test for this, I would implement these four test cases.
> 
> Is that what you were alluding to, @efb4f5ff-1298-471a-8973-3d47447115dc ?

Sorry i had to provide examples. From looking at your cases they look good. Everyone in our team does it a bit different depending on what type of PR they're sending in but these examples should give you a good view with what i meant with testcases. 

- https://github.com/FreeTubeApp/FreeTube/pull/5743
- https://github.com/FreeTubeApp/FreeTube/pull/7349
- https://github.com/FreeTubeApp/FreeTube/pull/5125
- https://github.com/FreeTubeApp/FreeTube/pull/7338

efb4f5ff-1298-471a-8973-3d47447115dc commented:
Also just a tip make sure your branch is up to date with the latest changes from the development branch before opening your PR.

efb4f5ff-1298-471a-8973-3d47447115dc reviewed (changes requested):
1. Sort by `Most Popular`
2. Mark a few videos as Watched
3. Enable Hide Watched
4. See that it doesnt hide the videos
5. Disable Hide Watched
6. See that it does hide the videos


https://github.com/user-attachments/assets/822e8229-6403-4419-86f8-d1d35f008774

palharesf commented:
Ok, I updated the 'Testing' section of the PR to reflect the examples provided, thank you for those.

I did run `git fetch origin` and `git merge origin/development` before running `git push origin feature/4497-hide-watched-videos` -- all the console messages I got indicated my branch was up-to-date with the dev branch, so I'm sorry if that was not the case.

And the case you mentioned (i.e. marking videos as watched and having them only hidden in the next refresh) was mentioned in my PR as something intentional - I decided against refreshing the video Panel every time one video was marked as watched because I considered it excessive. That said, I can work on an event watcher that tracks videos being marked as 'Watched' manually and forcing a refresh of the Video Panel if that's the preferred way moving forward. I'll include that in my next batch of changes along with the Live and Shorts pages. Thank you!

efb4f5ff-1298-471a-8973-3d47447115dc commented:
> And the case you mentioned (i.e. marking videos as watched and having them only hidden in the next refresh) was mentioned in my PR as something intentional - I decided against refreshing the video Panel every time one video was marked as watched because I considered it excessive. That said, I can work on an event watcher that tracks videos being marked as 'Watched' manually and forcing a refresh of the Video Panel if that's the preferred way moving forward. 

I think you interpreted my comment wrong. I pointed out a bug. Look at the screen recording again and try the steps to reproduce for yourself. It hides the videos when the toggle is OFF and doesnt hide them when the toggle is ON

palharesf commented:
> > And the case you mentioned (i.e. marking videos as watched and having them only hidden in the next refresh) was mentioned in my PR as something intentional - I decided against refreshing the video Panel every time one video was marked as watched because I considered it excessive. That said, I can work on an event watcher that tracks videos being marked as 'Watched' manually and forcing a refresh of the Video Panel if that's the preferred way moving forward.
> 
> I think you interpreted my comment wrong. I pointed out a bug. Look at the screen recording again and try the steps to reproduce for yourself. It hides the videos when the toggle is OFF and doesnt hide them when the toggle is ON

You're right, I had misunderstood your request. I apologize. I believe forcing the video panel to refresh once a video is marked as watched would fix this issue.

I've been trying to implement that, but the code for marking as watched is in 'ft-list-video.js' and the code for refreshing the video panel on the Channel page is, naturally, on Channel.vue. I don't want to tightly couple the two pages, so what's the preferred approach here? Using the Vuex store to manage state?

I don't have a lot of familiarity with Vuex, but I know how to create an event emitter and listener more generally and I assume that's the preferred route here. Any pointers would be welcome.

I'll implement the toggle on the Live and Shorts pages when this is working

absidue commented:
There are a few issues here but they are stem from the fact that you are overwriting the latestVideos array, which is why the setting is not working properly and also why you have to do the anti-pattern of fetching the data again from YouTube each time the switch is toggled.

So instead of overwriting the array and causing various problems, could instead create a computed which takes that array, checks the current value of the switch and if it is enabled returns a filtered copy of it, otherwise the original array. You can then reference that computed in the template instead of the latestVideos array. That way you don't have to do the unnecessary extra API requests and it will also automatically update when videos have been watched instead of only when you toggle the switch, it will also work properly with loading more data.

palharesf commented:
@absidue this is gold. I was not super familiar with the computed fields, but I implemented your suggestion and besides being simpler and cleaner, it works perfectly and is very fast.

I was also able to remove the code that forces a refresh of the videoPanel when the toggle is pressed, so everything looks much cleaner and responsive. Thanks a lot for this massive tip.

@efb4f5ff-1298-471a-8973-3d47447115dc I believe this also fixes the bug you encountered.

I'll implement the toggle for Live and Shorts before updating the PR, thank you all.

efb4f5ff-1298-471a-8973-3d47447115dc reviewed (commented):
These are some nice improvements.

I would remove the toggle when it changes sort order just like how it removes the dropdown. This gives it a bit of a cleaner look and feel

https://github.com/user-attachments/assets/27974a87-1e55-40f3-8bad-f7411bad8606

efb4f5ff-1298-471a-8973-3d47447115dc commented:
I tried to hide every video to see what would happen. The user must click on load more videos when all are currently loaded videos are marked as watched. Maybe its best to recognize when this is the case and auto load more videos. So if current window contains no videos because all are watched automatically load more. 


https://github.com/user-attachments/assets/5d217c62-bc1a-447e-90d8-e3fbbbcd6d8e

palharesf commented:
Really good tests, I appreciate it @efb4f5ff-1298-471a-8973-3d47447115dc . I have a decent idea on how to fix both, working on them now

palharesf commented:
Ok, I think my current changes fixes both of those issues. I recorded a screen cap that should show it:

![electron_enqcPX6ttY](https://github.com/user-attachments/assets/43de6dfe-2009-4a6a-a551-3965bcad5125)

The one thing I noticed is that, when I mark all the videos on a page as 'Watched', I get the following error as a toast:

`TypeError: Cannot read properties of null (reading 'getContinuation')`

It looks like the error is being caught in the `getChannelVideosLocalMore()` function but I can't seem to follow what's happening. I'll push the current changes but would appreciate some guidance on what's happening in this case and how to prevent it.

absidue commented:
You are likely calling the get more functions when there are no more continuations/you got to the end. The get more functions were not designed to handle being called when there is no continuation left, as previously they were only called in response to a user clicking the button and the button was hidden when there was nothing left.

Personally I would have just left it up to the user to manually press load more, but then again I'm not the target audience for this, so it won't be spamming requests to YouTube on my computer...

absidue commented on src/renderer/views/Channel/Channel.vue:
```diff
@@ -39,6 +39,14 @@
+        <FtToggleSwitch
+          v-if="(currentTab === 'videos') || (currentTab === 'shorts') || (currentTab === 'live')"
+          v-show="!isElementListLoading"
+          v-model="hideWatchedToggle"
```
FtToggleSwitch doesn't support v-model, you've not noticed it until now because you aren't actually tracking the value of the switch.
```suggestion
          :default-value="hideWatchedToggle"
```

absidue commented on src/renderer/views/Channel/Channel.vue:
```diff
@@ -341,6 +350,12 @@ const isElementListLoading = ref(false)
 
+const hideWatchedToggle = ref(false)
+
+const updateHideWatchedToggle = () => {
+  hideWatchedToggle.value = !hideWatchedToggle.value
+}
```
```suggestion
/**
 * @param {boolean} value
 */
function updateHideWatchedToggle(value) {
  hideWatchedToggle.value = value
}
```

absidue commented on src/renderer/views/Channel/Channel.vue:
```diff
@@ -1048,6 +1063,34 @@ const videoContinuationData = shallowRef(null)
 const showVideoSortBy = ref(true)
 const videoSortBy = ref('newest')
 
+const filteredVideos = computed(() => {
+  const filteredArray = hideWatchedToggle.value ? filterWatchedArray(latestVideos.value) : latestVideos.value
+  if (hideWatchedToggle.value && filteredArray.length === 0) {
+    if (process.env.SUPPORTS_LOCAL_API && apiUsed === 'local') {
+      getChannelVideosLocalMore()
+    } else {
+      channelInvidiousVideos()
+    }
+  }
+  return filteredArray
```
Computeds should not have side-effects, please remove the fetch more stuff out if the computed.

From the Vue 3 docs but it also applies to Vue 2: https://vuejs.org/guide/essentials/computed.html#getters-should-be-side-effect-free

absidue commented on src/renderer/views/Channel/Channel.vue:
```diff
@@ -2315,6 +2358,15 @@ function handleSubscription() {
+
+function filterWatchedArray(videos) {
+  const historyCache = store.getters.getHistoryCacheById
+  return videos.filter(video => !historyEntryExists(video.videoId, historyCache))
+}
+
+function historyEntryExists(videoId, historyCache) {
+  return typeof historyCache[videoId] !== 'undefined'
+}
```
```suggestion
  return videos.filter(video => !Object.hasOwn(historyCache, video.videoId))
}
```

efb4f5ff-1298-471a-8973-3d47447115dc reviewed (commented):
two things i just want to point out. 

1. If only one video is marked as unwatched you should hide the dropdown because there is no point in sorting one video.

https://github.com/user-attachments/assets/1a953d98-e99f-402c-aa94-fd844085e2aa

2. Here you see me marking an entire page as watched which prompted it to load more videos. It loaded allot of videos but there is also no page after this one, this is the last page it can load. If I would mark the entire page as Watched it shouldnt try to load more videos because there isnt anything to load (It could be that you already accounted for this but i wanted to point it out just in case)


https://github.com/user-attachments/assets/32d91473-8e42-4eff-b785-3343ed79086d

palharesf commented:
I updated the PR, and now fetching more videos won't happen automatically. I think this is a good compromise and avoids unnecessary issues that were being brought up by the automatic fetching of additional videos.

I'll take a look later at hiding the 'Sort By' selector once there's only one video available in the video panel. I think I know how to implement it. Here hoping that after this, we're ready to put a lid on it. I'm sure there are several additional edge cases we could account for, but I'm starting to worry we're over-complicating something that should be simpler. We can always improve in the future but what's already coded more than resolve the original issue I came to support with.

efb4f5ff-1298-471a-8973-3d47447115dc reviewed (approved):
Thanks @palharesf for creating this PR and making the necessary changes so this feature will have the same standards as other features within the application which in turn creates a coherent UX. I hope we'll see more of your contributions in the future!

PikachuEXE commented on src/renderer/views/Channel/Channel.vue:
```diff
@@ -39,9 +39,17 @@
         <FtSelect
           v-if="showVideoSortBy"
-          v-show="currentTab === 'videos' && latestVideos.length > 0"
+          v-show="(currentTab === 'videos' && filteredVideos.length > 0 && !hideWatchedToggle) || (currentTab === 'videos' && filteredVideos.length > 1 && hideWatchedToggle)"
```
A bit too long
```suggestion
          v-show="currentTab === 'videos' && (hideWatchedToggle ? filteredVideos.length > 1 : filteredVideos.length > 0)"
```

PikachuEXE commented on src/renderer/views/Channel/Channel.vue:
```diff
@@ -51,7 +59,7 @@
         <FtSelect
           v-if="!hideChannelShorts && showShortSortBy"
-          v-show="currentTab === 'shorts' && latestShorts.length > 0"
+          v-show="(currentTab === 'shorts' && filteredShorts.length > 0 && !hideWatchedToggle) || (currentTab === 'shorts' && filteredShorts.length > 1 && hideWatchedToggle)"
```
similar to above

palharesf replied:
Done - commit https://github.com/FreeTubeApp/FreeTube/pull/7366/commits/6ad9f5b1c65e4d47d952c9ff4982c9224bd30af4

PikachuEXE commented on src/renderer/views/Channel/Channel.vue:
```diff
@@ -61,7 +69,7 @@
         <FtSelect
           v-if="!hideLiveStreams && showLiveSortBy"
-          v-show="currentTab === 'live' && latestLive.length > 0"
+          v-show="(currentTab === 'live' && filteredLive.length > 0 && !hideWatchedToggle) || (currentTab === 'live' && filteredLive.length > 1 && hideWatchedToggle)"
```
similar to above

palharesf replied:
Done - commit https://github.com/FreeTubeApp/FreeTube/pull/7366/commits/76aaf13b973894da42dac7354bd5eaf429f03469

PikachuEXE commented on src/renderer/views/Channel/Channel.vue:
```diff
@@ -341,6 +350,15 @@ const isElementListLoading = ref(false)
 const isSearchTabLoading = ref(false)
 const currentTab = ref('videos')
 
+const hideWatchedToggle = ref(false)
```
It's not storing an element but the value
```suggestion
const hideWatchedEntries = ref(false)
```

palharesf replied:
Done - commit https://github.com/FreeTubeApp/FreeTube/pull/7366/commits/59e6b1a8205ae0179873b206350f665e6f0292ae

PikachuEXE commented on src/renderer/views/Channel/Channel.vue:
```diff
@@ -341,6 +350,15 @@ const isElementListLoading = ref(false)
+/**
+ * @param {boolean} value
+ */
+function updatehideWatchedEntries(value) {
```
Find and replace mistake made with case insensitive mode I can see :)
```suggestion
function updateHideWatchedEntries(value) {
```

palharesf replied:
Good eye, I even replaced entries one by one to try to avoid stuff like that and it still escaped me. Just fixed it, thank you very much

palharesf commented:
> Thanks @palharesf for creating this PR and making the necessary changes so this feature will have the same standards as other features within the application which in turn creates a coherent UX. I hope we'll see more of your contributions in the future!

My pleasure! I'd be happy to keep an eye out for features involving the Channel page, as this is the page I've got somewhat familiar with. If there are any issues any of you think I could get involved, feel free to @ me :)

efb4f5ff-1298-471a-8973-3d47447115dc commented:
Hi @palharesf 

After reviewing the PR, I think we made the wrong call in approving the direction for the "Hide Watched Videos" toggle on the Channel page. I’ve had some time to think it over, and I feel it would be better to avoid adding new settings outside of the Settings page.

Settings like this should really be centralized in the Settings page to prevent any confusion or the potential of other settings being added to random pages. This is something we've been consistent about, so I believe it makes more sense to maintain that approach.

What we should do instead is reuse the existing setting that’s listed in the Subscription settings and move it into the General section on the Distraction Free page. This will help keep the interface clean and consistent with our usual approach.

Let me know your thoughts on this and if you need help with the change.

palharesf commented:
> Hi @palharesf
> 
> After reviewing the PR, I think we made the wrong call in approving the direction for the "Hide Watched Videos" toggle on the Channel page. I’ve had some time to think it over, and I feel it would be better to avoid adding new settings outside of the Settings page.
> 
> Settings like this should really be centralized in the Settings page to prevent any confusion or the potential of other settings being added to random pages. This is something we've been consistent about, so I believe it makes more sense to maintain that approach.
> 
> What we should do instead is reuse the existing setting that’s listed in the Subscription settings and move it into the General section on the Distraction Free page. This will help keep the interface clean and consistent with our usual approach.
> 
> Let me know your thoughts on this and if you need help with the change.

Hi @efb4f5ff-1298-471a-8973-3d47447115dc,
Conceptually, I'm aligned with the direction of centralizing all settings on the Settings page instead of having custom settings applied into / managed by different pages.

I just want to re-hash what I see as the proposed implementation in your vision then (assuming we're starting from my edited branch):

1. Remove my toggle from the Channels Page
2. Add a toggle in the Settings page (under "Distraction Free", "General", titled 'Hide Watched Videos')
3. Update localization settings (to reflect the changes above)
4. Link all other functions and logic from the current Channels page toggle to the new Settings toggle

In summary, is that it?
I think the hardest part for me will be familiarizing myself with Vuex, but it's a good time as any to become familiar with it. I just want to make sure the direction is good before moving ahead.

Also, should I open a new PR or simply implement the changes over this PR / branch? I can also change the main post to reflect the updated approach. Let me know

efb4f5ff-1298-471a-8973-3d47447115dc commented:
>     1. Remove my toggle from the Channels Page
> 
>     2. Add a toggle in the Settings page (under "Distraction Free", "General", titled 'Hide Watched Videos')
> 
>     3. Update localization settings (to reflect the changes above)
> 
>     4. Link all other functions and logic from the current Channels page toggle to the new Settings toggle
> 
> 
> In summary, is that it? I think the hardest part for me will be familiarizing myself with Vuex, but it's a good time as any to become familiar with it. I just want to make sure the direction is good before moving ahead.

1. Correct
2. Correct but we can use the title of the already existing toggle that resides withing the Subscription Settings `Hide Videos on Watch`
3. See 2, so that would mean you can remove your localization and use the existing one. The placement does need to be changed so from https://github.com/FreeTubeApp/FreeTube/blob/development/static/locales/en-US.yaml#L531 to reside within the distraction free settings keys https://github.com/FreeTubeApp/FreeTube/blob/development/static/locales/en-US.yaml#L537
4. Correct, in addition to that remove the Subscription Settings `Hide Videos on Watch` toggle and add that functionality to the toggle you will add within the Distraction Free Settings. So that toggle will Hide Videos from the subscription page and the channel page.
5. Now that it will reside within the General section of the Distraction Free Settings I think it would also be good to add a tooltip with the description that this only applies to those two pages.


> 
> Also, should I open a new PR or simply implement the changes over this PR / branch? I can also change the main post to reflect the updated approach. Let me know

Feel free to take the approach you feel comfortable with

absidue commented:
Just to clarify you do not need to add a new setting, you would just need to move the existing distraction free setting around in the same file and then change your channel page logic to use that existing setting instead of its own switch.

palharesf commented:
Ok, good clarifications. Thank you, I'll get started on it now

palharesf commented:
OK a few questions to ensure I'm on the right track:

1. It looks like settings.js is where all components get assembled together, but Distraction Free Settings and Subscription Settings are actually two different components. So I would move the 'Hide Videos on Watch' toggle from subscription-settings.vue to distraction-settings.vue according to what we previously discussed right?
2. I assume I don't need to make any changes to the store if I just move the toggle from one settings component to the other, correct?
3. From there I can just access the settings value from the store within the Channels page and update all the logic to reference it right?

I'll look into the tooltip as soon as I got the basic logic working properly. Thank you all!

palharesf commented:
I implemented the base functionality (still missing the tooltip, working on it now) -- but it's not evident to me where are the conflicts that the bot is warning me of. I'd appreciate some guidance on that while I get the tooltip up and running

efb4f5ff-1298-471a-8973-3d47447115dc commented on static/locales/en-US.yaml:
```diff
@@ -1086,6 +1086,7 @@ Tooltips:
        The channel ID entered must be a complete match and is case sensitive.
     Hide Subscriptions Live: 'This setting is overridden by the app-wide "{appWideSetting}" setting, in the "{subsection}" section of the "{settingsSection}"'
     Hide Videos and Playlists Containing Text: Enter a word, word fragment, or phrase (case insensitive) to hide all videos & playlists whose original titles contain it throughout all of FreeTube, excluding only History, Your Playlists, and videos inside of playlists.
+    Hide Videos on Watch: 'This option hides watched videos, lives and shorts from the Subscriptions and Channels Pages. It does not affect the Home tab on the Channels page"'
```
```suggestion
    Hide Videos on Watch: This option hides Watched Videos, Live streams, and Shorts from the Subscriptions and Channels pages. It does not affect the Home tab on a Channel’s page.
```

palharesf commented:
Ok I think we're in a good spot. Tooltip implemented (and wording reviewed), functionality seems to be working for all the cases we expected (in the subscriptions page it's unchanged, for the Channels page it works for videos, lives and shorts), and the functionality to hide the Sorting button on the Channels page when there's only one video being displayed is still working. I still didn't understand the conflicts generated by GitHub Actions, but happy to resolve them with some guidance.

Please let me know if there's any additional work anyone notices that I missed. Otherwise, thank you all for the help and I'll start looking at related issues to tackle once we close this :)

efb4f5ff-1298-471a-8973-3d47447115dc commented:
We're all a bit busy at the moment. We will get back to you as soon as possible. I cant put my finger on why you are receiving the conflicts. What you can try is updating your branch so your changes are being made on top of the newest version of the dev branch

PikachuEXE commented:
The conflict is caused by migration to composition API https://github.com/FreeTubeApp/FreeTube/pull/7335
Some files edited in this PR (before merge) no longer exist
I did the merge for you since it's rare

PikachuEXE reviewed (commented):
The screenshots might need to be updated for later reviewers...

palharesf commented:
Thank you for addressing the conflicts @PikachuEXE. 

> The screenshots might need to be updated for later reviewers... `reload video list once a video is mark as 'Watched' manually` is done already too right?

Since we started using the computed property `filteredVideos` in the videoPanel, we don't need to reload the panel every time a video is watched. This is done automatically. I forgot to update the main PR post, I'll go back to it tomorrow and update the screenshots and tests. Thank you

efb4f5ff-1298-471a-8973-3d47447115dc commented on static/locales/en-US.yaml:
```diff
@@ -1086,6 +1086,7 @@ Tooltips:
        The channel ID entered must be a complete match and is case sensitive.
     Hide Subscriptions Live: 'This setting is overridden by the app-wide "{appWideSetting}" setting, in the "{subsection}" section of the "{settingsSection}"'
     Hide Videos and Playlists Containing Text: Enter a word, word fragment, or phrase (case insensitive) to hide all videos & playlists whose original titles contain it throughout all of FreeTube, excluding only History, Your Playlists, and videos inside of playlists.
+    Hide Videos on Watch: 'This option hides watched videos, lives and shorts from the Subscriptions and Channels Pages. It does not affect the Home tab on the Channels page"'
```
I reread this again and thought it wasnt very clear

```suggestion
    Hide Videos on Watch: 'Hides watched videos from the Videos, Shorts, and Live tabs on the Subscription and Channel pages. This does not affect the Home tab on Channel pages'
```

efb4f5ff-1298-471a-8973-3d47447115dc replied:
This one needs to be commited again

palharesf commented:
Ok, I updated the main post with the current implementation, current screenshots and explanations

While doing some testing I noticed the logic for hiding the 'Sort By' dropdown when there was only one live left in the `videoPanel` was referencing to the `filteredShorts` variable instead of the `filteredLive` variable, so it was not working. I fixed that, but when I pushed my changes I think I undid the fix for conflicts that PikachuEXE had done -- apologies for that

palharesf commented:
Hi @absidue,

Sounds good, and sorry again. I had force-pushed yesterday without apparent negative consequences, and wrongly assumed I could do the same again without realizing it would undo the fixes, but I recognize my mistake. Are we able to revert my last push, so we return to the previous state (without conflicts)? There is only one line changed between that state and the current and might be an easier fix than addressing the conflicts again.

Since we're already at it, was there anything I could have done besides what I outlined here (https://github.com/FreeTubeApp/FreeTube/pull/7366#issuecomment-2909670290) to fix my branch?

palharesf commented:
One change that is still required (but that I've been unable to push) is on 'Channels.vue', line 64. As far as I can see it still has `hideWatchedSubs ? filteredShorts.length > 1 : filteredShorts.length > 0`, where it should be `hideWatchedSubs ? filteredLive.length > 1 : filteredLive.length > 0`, but I am unable to push those changes from my side.

This was the fix that I had force-pushed previously and that broke the fix for the conflicts

PikachuEXE commented on src/renderer/views/Channel/Channel.vue:
```diff
@@ -41,7 +41,7 @@
         <FtSelect
           v-if="showVideoSortBy"
-          v-show="currentTab === 'videos' && latestVideos.length > 0"
+          v-show="currentTab === 'videos' && (hideWatchedSubs ? filteredVideos.length > 1 : filteredVideos.length > 0)"
```
No idea why we need different logic based on `Hide Videos on Watch` enabled or not
Probably just need `filteredVideos.length > 1` here

Or do you mean the opposite?
When watched entries are hidden, allow changing the sorting due to having 1 entry does not mean 1 entry only in other orders (e.g. all except 1 watched, but all oldest still not watched)
However if this is the case the criteria should be on NOT filtered entries = `latestVideos.length > 1` (i.e. more than 1 entries from remote = possibility to have unwatched entries in other sorting orders

I might get some of this wrong let me know

absidue replied:
The original code checks YouTube's response for the dropdown, which only hides it when there is only 1 or 0 videos and there are no continuations. You need to ensure that you do not break this behaviour when hide watched is disabled.

So the important thing for this pull requests is that you cannot hide it just based on filteredVideos.length < 2, because if there is a continuation available, you don't actually know that all videos are hidden, because there may be some visible ones that were not fetched yet.

palharesf replied:
My proposal was to only show the dropdown if we are in one of the tabs it would appear (`currentTab === 'videos'` and the same for shorts and lives) and if the array containing the videos to be displayed, after computing for hiding watched videos, has more than one video (`filteredVideos.length >1)

So basically just switching 
` v-show="currentTab === 'videos' && (hideWatchedSubs ? filteredVideos.length > 1 : filteredVideos.length > 0)"`

for
` v-show="currentTab === 'videos' && filteredVideos.length > 1"

It makes sense to me unless I'm missing something, and I just tested and it works. Let me know if you see any issues with that approach

palharesf replied:
I also realized why I was not able to push my previous changes cleanly, my forked repo was behind the actual origin repo so that's why simply pulling the changes was not working. I have just synced my repo with the origin one, and when I try to pull the changes into my repo I finally get the conflicts. 

I assume that I can accept the remote repository changes by using `git checkout --theirs src/renderer/components/subscription-settings/subscription-settings.js` and `git checkout --theirs src/renderer/components/subscription-settings/subscription-settings.vue`, but I'm afraid of doing that without checking with you all and breaking things. If its easier to just suggest the changes I outlined in my previous comment directly into the code, I'm also happy with that approach

palharesf replied:
Hi, I still haven't been able to properly merge the changes from the remote branch onto my local copy of the repo, as outlined [here](https://github.com/FreeTubeApp/FreeTube/pull/7366#discussion_r2113912398).

It's unclear what I can do without breaking things, and as I have already broken stuff here in the past I decided to wait for guidance.

Ein-Tim replied:
I don't know your needs, but if you prefer a GUI over the command line, I recommend using GitHub desktop: https://github.com/apps/desktop

palharesf replied:
> The original code checks YouTube's response for the dropdown, which only hides it when there is only 1 or 0 videos and there are no continuations. You need to ensure that the original behaviour still exists when the hide watched setting is disabled.
> 
> So the important thing for this pull requests is that you cannot hide it just based on filteredVideos.length < 2, because if there is a continuation available, you don't actually know that all videos are hidden, because there may be some visible ones that were not fetched yet.

@absidue I am reviewing the code now, and since `filteredVideos` is computed from `latestVideos`, I would assume it preserves its properties i.e. if `latestVideos` had no continuations, `filteredVideos` will have no continuations. Is that reasonable?

As a way of testing that, I logged the length of both `filteredVideos` and `latestVideos` for a channel where I have hidden two videos from the first page. You can see that the offset is always two, i.e. even initially latestVideos does not hold data from the whole playlist, just from the first page. While I can try to incorporate data from the full playlist (including continuations) in the current check, it doesn't look like this is how it works currently with `latestVideos` unless I'm missing something here.

![image](https://github.com/user-attachments/assets/2d8c2085-777d-422a-a290-f31269a8fd8d)

TL;DR - if the current implementation was taking continuations into account, I believe `latestVideos` should be initialized with 117 instead of 30. Let me know if I'm not understanding something.

palharesf replied:
Assuming we agree on our point regarding continuations (i.e. my changes do not modify the base behavior), I wanted to push the last remaining commit but I'm getting an eslint error.

Changes I'm trying to commit:

![image](https://github.com/user-attachments/assets/f5a0fc3c-3081-441b-9d36-228f83af6eb5)

Error message I'm getting (even after creating a new branch and checking out this PR):

`> git -c user.useConfigOnly=true commit --quiet --allow-empty-message --file -
╭────────────────────────────────────────╮
│ 🥊 lefthook v1.11.10  hook: pre-commit │
╰────────────────────────────────────────╯
│  stylelint (skip) no files for inspection
! The local project doesn't define a 'packageManager' field. Corepack will now add one referencing yarn@1.22.22+sha512.a6b2f7906b721bba3d67d4aff083df04dad64c399707841b7acf00f6b133b7ac24255f2652fa22ae3534329dc6180534e98d17432037ff6fd140556e2bb3137e.
! For more details about this field, consult the documentation at https://nodejs.org/api/packages.html#packagemanager

(node:12344) ExperimentalWarning: Importing JSON modules is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)

Oops! Something went wrong! :(

ESLint: 9.25.0

TypeError: Key "rules": Key "unicorn/prefer-single-call": Could not find "prefer-single-call" in plugin "unicorn".
    at throwRuleNotFoundError (C:\Users\palha\Documents\GitHub\FreeTube\node_modules\eslint\lib\config\rule-validator.js:66:16)
    at RuleValidator.validate (C:\Users\palha\Documents\GitHub\FreeTube\node_modules\eslint\lib\config\rule-validator.js:149:5)
    at new Config (C:\Users\palha\Documents\GitHub\FreeTube\node_modules\eslint\lib\config\config.js:255:18)
    at [finalizeConfig] (C:\Users\palha\Documents\GitHub\FreeTube\node_modules\eslint\lib\config\flat-config-array.js:212:10)
    at FlatConfigArray.getConfigWithStatus (C:\Users\palha\Documents\GitHub\FreeTube\node_modules\@eslint\config-array\dist\cjs\index.cjs:1263:55)
    at FlatConfigArray.getConfig (C:\Users\palha\Documents\GitHub\FreeTube\node_modules\@eslint\config-array\dist\cjs\index.cjs:1281:15)
    at C:\Users\palha\Documents\GitHub\FreeTube\node_modules\eslint\lib\eslint\eslint.js:747:28
    at async Promise.all (index 0)
    at async ESLint.lintFiles (C:\Users\palha\Documents\GitHub\FreeTube\node_modules\eslint\lib\eslint\eslint.js:743:19)
    at async Object.execute (C:\Users\palha\Documents\GitHub\FreeTube\node_modules\eslint\lib\cli.js:638:14)
error Command failed with exit code 2.
┃  eslint ❯ 

yarn run v1.22.22
$ C:\Users\palha\Documents\GitHub\FreeTube\node_modules\.bin\eslint --no-color src/renderer/views/Channel/Channel.vue
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

exit status 2                                      
  ────────────────────────────────────
summary: (done in 7.21 seconds)       
🥊 eslint
`

I believe these changes would resolve the outstanding comments on this PR while addressing the concerns raised on the discussion here, but I'm unsure on how to push them with this new error. If it's easier, someone could suggest those changes and I could simple approve them on this PR, otherwise help on how to solve this issue locally would be welcome. Thanks

efb4f5ff-1298-471a-8973-3d47447115dc replied:
You can try do request changes yourself on the lines you want to change. After the change is submitted a button should appear to commit the change.

https://github.com/user-attachments/assets/6124c64b-b598-4308-8c0e-88474ac938f2

PikachuEXE commented on src/renderer/views/Channel/Channel.vue:
```diff
@@ -51,7 +51,7 @@
         <FtSelect
           v-if="!hideChannelShorts && showShortSortBy"
-          v-show="currentTab === 'shorts' && latestShorts.length > 0"
+          v-show="currentTab === 'shorts' && (hideWatchedSubs ? filteredShorts.length > 1 : filteredShorts.length > 0)"
```
ditto

PikachuEXE commented on src/renderer/views/Channel/Channel.vue:
```diff
@@ -61,7 +61,7 @@
         <FtSelect
           v-if="!hideLiveStreams && showLiveSortBy"
-          v-show="currentTab === 'live' && latestLive.length > 0"
+          v-show="currentTab === 'live' && (hideWatchedSubs ? filteredLive.length > 1 : filteredLive.length > 0)"
```
ditto

palharesf commented:
@efb4f5ff-1298-471a-8973-3d47447115dc I still don't understand the issue with continuations. As per my comment here (https://github.com/FreeTubeApp/FreeTube/pull/7366#discussion_r2132754590), it seems to me the original array didn't check for continuation for deciding when to show the "Fetch More" button, so I kept the same logic for the computed array. If we want to change that logic I think it's fine, but it seems to me it's a different issue from hiding watched videos altogether. If I'm wrong, I'd appreciate understanding how, since from my current point of view that doesn't seem like an issue. I'm less stuck on code and more stuck on comprehension here.

Regarding the dropdown, it's still visible when there's only one video being displayed on the channel because of the last change made by PikachuEXE (https://github.com/FreeTubeApp/FreeTube/pull/7366/commits/b232adf1ceec138c1c69eb44d01eceacc2a98c11). I tested this just now and this is happening because the array responsible for displaying videos on the screen is filteredVideos, whereas the array responsible for displaying the Sort element is latestVideos. They don't match when we have the 'Hide Videos on Watch' setting enabled, so naturally you'll have situations like this one, where the size of the arrays differ and hence the logic doesn't work as intended.

![image](https://github.com/user-attachments/assets/a5b14f58-b30c-458b-b9af-77690a764a86)

My proposed solution here is to revert and let filteredVideos command whether the Sort element would be displayed or not, but naturally I'd like @PikachuEXE two cents here since he's the one that changed this bit of logic and can show me the shortcomings of my suggestion.



Apologies if it seemed I was MIA, but I honestly did not know you all were waiting on me for the pending changes and I already forgot most of what I worked on in this PR since I dove into it a while ago.

efb4f5ff-1298-471a-8973-3d47447115dc commented:
@palharesf thank you so much for you contribution and bearing with us. Normally merging a PR doesn't take this long. Let me know if you want to work on another issue and i can help you find one that fits your criteria
